crates/prune/src/pruner.rs

```diff
@@ -253,13 +263,149 @@ impl<DB: Database> Pruner<DB> {
                     "Pruned receipts"
                 );
             },
+            |_| false,
         )?;
 
         provider.save_prune_checkpoint(
             PrunePart::Receipts,
             PruneCheckpoint { block_number: to_block, prune_mode },
         )?;
 
+        // `PrunePart::Receipts` overrides `PrunePart::ContractLogs`, so we can preemptively
+        // limit their pruning start point.
+        provider.save_prune_checkpoint(
+            PrunePart::ContractLogs,
+            PruneCheckpoint { block_number: to_block, prune_mode },
+        )?;
+
+        Ok(())
+    }
+
+    /// Prune receipts up to the provided block by filtering logs. Works as in inclusion list, and
+    /// removes every receipt not belonging to it.
+    #[instrument(level = "trace", skip(self, provider), target = "pruner")]
+    fn prune_receipts_by_logs(
+        &self,
+        provider: &DatabaseProviderRW<'_, DB>,
+        tip_block_number: BlockNumber,
+    ) -> PrunerResult {
+        // Contract log filtering removes every receipt possible except the ones in the list. So,
+        // for the other receipts it's as if they had a `PruneMode::Distance()` of 128.
+        let to_block = PruneMode::Distance(MINIMUM_PRUNING_DISTANCE)
+            .prune_target_block(
+                tip_block_number,
+                MINIMUM_PRUNING_DISTANCE,
+                PrunePart::ContractLogs,
+            )?
+            .map(|(bn, _)| bn)
+            .unwrap_or_default();
+
+        // Figure out what receipts have already been pruned, so we can have an accurate
+        // `address_filter`
+        let pruned = provider
+            .get_prune_checkpoint(PrunePart::ContractLogs)?
+            .map(|checkpoint| checkpoint.block_number)
+            .unwrap_or(0);
+
+        let address_filter =
+            self.modes.contract_logs_filter.group_by_block(tip_block_number, Some(pruned))?;
+
+        // Split all transactions in different block ranges. Each block range will have its own
+        // filter address list.
+        let mut block_ranges = vec![];
+        let mut blocks_iter = address_filter.keys().peekable();
+        let mut filtered_addresses = vec![];
+
+        while let Some(start_block) = blocks_iter.next() {
+            filtered_addresses.extend_from_slice(address_filter.get(start_block).expect("qed"));
+
+            // This will clear all receipts before the first  appearance of a contract log
+            if block_ranges.is_empty() {
+                block_ranges.push((0, *start_block, 0));
+            }
+
+            let end_block =
+                blocks_iter.peek().map(|next_block| *next_block - 1).unwrap_or(to_block);
```

Comment:
we do include the target block, since `get_next_tx_num_range_from_checkpoint` uses an inclusive range... so it becomes something like

```
0 .. =(end_block_1-1)
end_block_1..=(end_block_2-1)
end_block_2..=to_block
```